File: pick_and_place/src/pick_place.cpp
#include <ros/ros.h>

// MoveIt
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <geometry_msgs/Pose.h>
#include <tf/transform_listener.h>

#include "tf/transform_datatypes.h"

// TF2
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>

void openGripper(trajectory_msgs::JointTrajectory& posture)
{
  posture.joint_names.resize(6);
  posture.joint_names[0] = "gripper_left_driver_joint";
  posture.joint_names[1] = "gripper_left_follower_joint";
  posture.joint_names[2] = "gripper_left_spring_link_joint";
  posture.joint_names[3] = "gripper_right_driver_joint";
  posture.joint_names[4] = "gripper_right_follower_joint";
  posture.joint_names[5] = "gripper_right_spring_link_joint";

  posture.points.resize(1);
  posture.points[0].positions.resize(6);
  posture.points[0].positions[0] = 0;
  posture.points[0].positions[1] = 0;
  posture.points[0].positions[2] = 0;
  posture.points[0].positions[3] = 0;
  posture.points[0].positions[4] = 0;
  posture.points[0].positions[5] = 0;
  posture.points[0].time_from_start = ros::Duration(0.5);
}

void closedGripper(trajectory_msgs::JointTrajectory& posture)
{
posture.joint_names.resize(6);
  posture.joint_names[0] = "gripper_left_driver_joint";
  posture.joint_names[1] = "gripper_left_follower_joint";
  posture.joint_names[2] = "gripper_left_spring_link_joint";
  posture.joint_names[3] = "gripper_right_driver_joint";
  posture.joint_names[4] = "gripper_right_follower_joint";
  posture.joint_names[5] = "gripper_right_spring_link_joint";

  posture.points.resize(1);
  posture.points[0].positions.resize(6);
  posture.points[0].positions[0] = 0;
  posture.points[0].positions[1] = 0;
  posture.points[0].positions[2] = 0;
  posture.points[0].positions[3] = 0.4;
  posture.points[0].positions[4] = 0;
  posture.points[0].positions[5] = 0;
  posture.points[0].time_from_start = ros::Duration(0.5);
}

void pick(moveit::planning_interface::MoveGroupInterface& move_group,const geometry_msgs::Pose &marker_pose)
{
  // 创建一个抓取向量，这里只创建一个抓取
  std::vector<moveit_msgs::Grasp> grasps;
  grasps.resize(1);

  // 设置抓取姿态

  tf2::Quaternion orientation;

  if(marker_pose.position.z>0.1){

    grasps[0].grasp_pose.header.frame_id = "ar_marker_1";
    orientation.setRPY(M_PI/2, 0, M_PI);
    grasps[0].grasp_pose.pose.orientation = tf2::toMsg(orientation);
    grasps[0].grasp_pose.pose.position.x = 0;
    grasps[0].grasp_pose.pose.position.y = -0.15;
    grasps[0].grasp_pose.pose.position.z =-(marker_pose.position.z+0.05)/2.0;

    // 设置 pre-grasp approach
    grasps[0].pre_grasp_approach.direction.header.frame_id = "ar_marker_1";

    grasps[0].pre_grasp_approach.direction.vector.y = 1.0;
    grasps[0].pre_grasp_approach.min_distance = 0.095;
    grasps[0].pre_grasp_approach.desired_distance = 0.115;

  }
  else{
    grasps[0].grasp_pose.header.frame_id = "ar_marker_0";
    orientation.setRPY(M_PI, 0, 0);
    grasps[0].grasp_pose.pose.orientation = tf2::toMsg(orientation);
    grasps[0].grasp_pose.pose.position.x = 0;
    grasps[0].grasp_pose.pose.position.y = 0;
    grasps[0].grasp_pose.pose.position.z = 0.13;

    // 设置 pre-grasp approach
    grasps[0].pre_grasp_approach.direction.header.frame_id = "base_link";

    grasps[0].pre_grasp_approach.direction.vector.z = -1.0;
    grasps[0].pre_grasp_approach.min_distance = 0.095;
    grasps[0].pre_grasp_approach.desired_distance = 0.115;

  }

  // 设置 post-grasp retreat
  grasps[0].post_grasp_retreat.direction.header.frame_id = "base_link";

  grasps[0].post_grasp_retreat.direction.vector.z = 1.0;
  grasps[0].post_grasp_retreat.min_distance = 0.1;
  grasps[0].post_grasp_retreat.desired_distance = 0.25;

  // 设置抓取前爪子的状态
  openGripper(grasps[0].pre_grasp_posture);

  // 设置爪子抓取的状态
  closedGripper(grasps[0].grasp_posture);

  // 设置支撑面.
  move_group.setSupportSurfaceName("table1");

  // 调用pick使用给定的grasps抓取对象
  if(marker_pose.position.z>0.1){
    move_group.pick("object1", grasps);
  }
  else{
    move_group.pick("object0", grasps);
  }

}

void place(moveit::planning_interface::MoveGroupInterface& group,const geometry_msgs::Pose &marker_pose)
{
  // 创建一个尝试的向量，这里只创建一个放置位置
  std::vector<moveit_msgs::PlaceLocation> place_location;
  place_location.resize(1);

  // 设置放置的姿态
  place_location[0].place_pose.header.frame_id = "base_link";
  tf2::Quaternion orientation;
  orientation.setRPY(0, 0, M_PI / 2);
  place_location[0].place_pose.pose.orientation = tf2::toMsg(orientation);

  //放置对象
  if(marker_pose.position.z>0.1){
    place_location[0].place_pose.pose.position.x = -0.5;
    place_location[0].place_pose.pose.position.y = 0.1;
    place_location[0].place_pose.pose.position.z = (marker_pose.position.z+0.05)/2.0-0.05;
  }
  else{
    place_location[0].place_pose.pose.position.x = -0.47;
    place_location[0].place_pose.pose.position.y = -0.1;
    place_location[0].place_pose.pose.position.z = (marker_pose.position.z+0.05)/2.0-0.05;
  }

  // 设置pre-place approach

  place_location[0].pre_place_approach.direction.header.frame_id = "base_link";

  place_location[0].pre_place_approach.direction.vector.z = -1.0;
  place_location[0].pre_place_approach.min_distance = 0.095;
  place_location[0].pre_place_approach.desired_distance = 0.115;

  // 设置 post-grasp retreat
  place_location[0].post_place_retreat.direction.header.frame_id = "base_link";
  if(marker_pose.position.z>0.1){
    place_location[0].post_place_retreat.direction.vector.x = 1.0;
    place_location[0].post_place_retreat.min_distance = 0.1;
    place_location[0].post_place_retreat.desired_distance = 0.15;
  }
  else{
    place_location[0].post_place_retreat.direction.vector.z = 1.0;
    place_location[0].post_place_retreat.min_distance = 0.08;
    place_location[0].post_place_retreat.desired_distance = 0.15;
  }

  //设置放置对象完成后机械爪的状态
  openGripper(place_location[0].post_place_posture);

  // 设置支撑面table2.
  group.setSupportSurfaceName("table2");

  // 调用place将对象放置到给定的位置
  if(marker_pose.position.z>0.1){
    group.place("object1", place_location);
  }
  else{
    group.place("object0", place_location);
  }
}

void addCollisionObjects(moveit::planning_interface::PlanningSceneInterface& planning_scene_interface,const geometry_msgs::Pose &marker0_pose,const geometry_msgs::Pose &marker1_pose)
{
  //先删除相关物体
  std::vector<std::string> object_ids;
  object_ids.push_back("object0");
  object_ids.push_back("object1");
  object_ids.push_back("table1");
  object_ids.push_back("table2");
  planning_scene_interface.removeCollisionObjects(object_ids);

  // 添加桌子和物块
  // 生成规划环境
  // 创建容纳2个碰撞对象的向量  
  std::vector<moveit_msgs::CollisionObject> collision_objects;
  collision_objects.resize(4);

  // 添加第一个放置ur机械臂的桌子.
  collision_objects[0].id = "table1";
  collision_objects[0].header.frame_id = "base_link";

  //定义桌子1的形状和尺寸
  collision_objects[0].primitives.resize(1);
  collision_objects[0].primitives[0].type = collision_objects[0].primitives[0].BOX;
  collision_objects[0].primitives[0].dimensions.resize(3);
  collision_objects[0].primitives[0].dimensions[0] = 0.4;
  collision_objects[0].primitives[0].dimensions[1] = 0.2;
  collision_objects[0].primitives[0].dimensions[2] = 0.4;

  //定义桌子1的位置
  collision_objects[0].primitive_poses.resize(1);
  collision_objects[0].primitive_poses[0].position.x = 0;
  collision_objects[0].primitive_poses[0].position.y = 0.5;
  collision_objects[0].primitive_poses[0].position.z = -0.25;


  collision_objects[0].operation = collision_objects[0].ADD;

  //添加第二张桌子
  collision_objects[1].id = "table2";
  collision_objects[1].header.frame_id = "base_link";

  //定义桌子的形状和尺寸
  collision_objects[1].primitives.resize(1);
  collision_objects[1].primitives[0].type = collision_objects[1].primitives[0].BOX;
  collision_objects[1].primitives[0].dimensions.resize(3);
  collision_objects[1].primitives[0].dimensions[0] = 0.2;
  collision_objects[1].primitives[0].dimensions[1] = 0.4;
  collision_objects[1].primitives[0].dimensions[2] = 0.4;

  //定义桌子的位置
  collision_objects[1].primitive_poses.resize(1);
  collision_objects[1].primitive_poses[0].position.x = -0.5;
  collision_objects[1].primitive_poses[0].position.y = 0;
  collision_objects[1].primitive_poses[0].position.z = -0.25;


  collision_objects[1].operation = collision_objects[1].ADD;

  // 定义将要抓取的物块0
  collision_objects[2].header.frame_id = "base_link";
  collision_objects[2].id = "object0";

  //定义物块0的形状和尺寸
  collision_objects[2].primitives.resize(1);
  collision_objects[2].primitives[0].type = collision_objects[2].primitives[0].BOX;
  collision_objects[2].primitives[0].dimensions.resize(3);
  collision_objects[2].primitives[0].dimensions[0] = 0.05;
  collision_objects[2].primitives[0].dimensions[1] = 0.05;
  collision_objects[2].primitives[0].dimensions[2] = marker0_pose.position.z+0.05;

  //定义物块0的位置
  collision_objects[2].primitive_poses.resize(1);
  collision_objects[2].primitive_poses[0].position.x = marker0_pose.position.x;
  collision_objects[2].primitive_poses[0].position.y = marker0_pose.position.y;
  collision_objects[2].primitive_poses[0].position.z = (marker0_pose.position.z+0.05)/2.0-0.05;
  collision_objects[2].primitive_poses[0].orientation.x = marker0_pose.orientation.x;
  collision_objects[2].primitive_poses[0].orientation.y = marker0_pose.orientation.y;
  collision_objects[2].primitive_poses[0].orientation.z = marker0_pose.orientation.z;
  collision_objects[2].primitive_poses[0].orientation.w = marker0_pose.orientation.w;


  collision_objects[2].operation = collision_objects[2].ADD;

    // 定义将要抓取的物块1
  collision_objects[3].header.frame_id = "base_link";
  collision_objects[3].id = "object1";

  //定义物块的形状和尺寸
  collision_objects[3].primitives.resize(1);
  collision_objects[3].primitives[0].type = collision_objects[3].primitives[0].BOX;
  collision_objects[3].primitives[0].dimensions.resize(3);
  collision_objects[3].primitives[0].dimensions[0] = 0.05;
  collision_objects[3].primitives[0].dimensions[1] = 0.05;
  collision_objects[3].primitives[0].dimensions[2] = marker1_pose.position.z+0.05;

  //定义物块的位置
  collision_objects[3].primitive_poses.resize(1);
  collision_objects[3].primitive_poses[0].position.x = marker1_pose.position.x;
  collision_objects[3].primitive_poses[0].position.y = marker1_pose.position.y;
  collision_objects[3].primitive_poses[0].position.z = (marker1_pose.position.z+0.05)/2.0-0.05;
  collision_objects[3].primitive_poses[0].orientation.x = marker1_pose.orientation.x;
  collision_objects[3].primitive_poses[0].orientation.y = marker1_pose.orientation.y;
  collision_objects[3].primitive_poses[0].orientation.z = marker1_pose.orientation.z;
  collision_objects[3].primitive_poses[0].orientation.w = marker1_pose.orientation.w;


  collision_objects[3].operation = collision_objects[3].ADD;

  planning_scene_interface.applyCollisionObjects(collision_objects);
}

int main(int argc, char** argv)
{
  ros::init(argc, argv, "pick_place_demo");
  ros::NodeHandle nh;
  ros::AsyncSpinner spinner(1);
  spinner.start();

  ros::WallDuration(1.0).sleep();
  moveit::planning_interface::PlanningSceneInterface planning_scene_interface;
  moveit::planning_interface::MoveGroupInterface group("arm");
  group.setPlanningTime(45.0);

  tf::TransformListener listener;
  tf::StampedTransform transform0;
  tf::StampedTransform transform1;
  while (ros::ok()) {
    try{
      listener.lookupTransform("base_link", "ar_marker_0",ros::Time(0), transform0);
      listener.waitForTransform("ar_marker_0", "tool0",ros::Time(0), ros::Duration(2));
      listener.lookupTransform("base_link", "ar_marker_1",ros::Time(0), transform1);
      listener.waitForTransform("ar_marker_1", "tool0",ros::Time(0), ros::Duration(2));

      break;
    }
    catch (tf::TransformException ex){
      ROS_INFO("Waiting for transform between 'base_link' and 'ar_marker_0'");
      ROS_INFO("Waiting for transform between 'ar_marker_0' and 'tool0'");
      ROS_INFO("Waiting for transform between 'base_link' and 'ar_marker_1'");
      ROS_INFO("Waiting for transform between 'ar_marker_1' and 'tool0'");
      ros::Duration(1.0).sleep();
    }
  }
  ROS_INFO("Found transform between 'base_link' and 'ar_marker_0'");
  ROS_INFO("Found transform between 'ar_marker_0' and 'tool0'");
  ROS_INFO("Found transform between 'base_link' and 'ar_marker_1'");
  ROS_INFO("Found transform between 'ar_marker_1' and 'tool0'");

  geometry_msgs::Pose marker0_pose;
  geometry_msgs::Pose marker1_pose;

  marker0_pose.position.x=transform0.getOrigin().x();
  marker0_pose.position.y=transform0.getOrigin().y();
  marker0_pose.position.z=transform0.getOrigin().z();
  marker0_pose.orientation.x=transform0.getRotation().getX();
  marker0_pose.orientation.y=transform0.getRotation().getY();
  marker0_pose.orientation.z=transform0.getRotation().getZ();
  marker0_pose.orientation.w=transform0.getRotation().getW();

  marker1_pose.position.x=transform1.getOrigin().x();
  marker1_pose.position.y=transform1.getOrigin().y();
  marker1_pose.position.z=transform1.getOrigin().z();
  marker1_pose.orientation.x=transform1.getRotation().getX();
  marker1_pose.orientation.y=transform1.getRotation().getY();
  marker1_pose.orientation.z=transform1.getRotation().getZ();
  marker1_pose.orientation.w=transform1.getRotation().getW();

  addCollisionObjects(planning_scene_interface,marker0_pose,marker1_pose);


  ros::WallDuration(1.0).sleep();

  pick(group,marker0_pose);

  ros::WallDuration(1.0).sleep();

  place(group,marker0_pose);

  ros::WallDuration(1.0).sleep();

  //group.setNamedTarget("up");
  //group.move();

  pick(group,marker1_pose);

  ros::WallDuration(1.0).sleep();

  place(group,marker1_pose);

  ros::WallDuration(1.0).sleep();

  group.setNamedTarget("home");
  group.move();

  ros::WallDuration(1.0).sleep();

	ros::shutdown(); 
  return 0;
}

File: pick_and_place/scripts/pick_place.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy, sys
import moveit_commander
import tf
from geometry_msgs.msg import PoseStamped, Pose
from moveit_commander import MoveGroupCommander, PlanningSceneInterface
from moveit_msgs.msg import PlanningScene, ObjectColor
from moveit_msgs.msg import Grasp, GripperTranslation, MoveItErrorCodes

from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
from tf.transformations import quaternion_from_euler
from copy import deepcopy

GROUP_NAME_ARM = 'arm'
GROUP_NAME_GRIPPER = 'gripper'

GRIPPER_FRAME = 'ee_link'

GRIPPER_OPEN = [0.04,0,0,0]
GRIPPER_CLOSED = [0.35]

REFERENCE_FRAME = 'base_link'

class PickAndPlaceDemo:
    def __init__(self):
        # 初始化move_group的API
        moveit_commander.roscpp_initialize(sys.argv)
        
        # 初始化ROS节点
        rospy.init_node('pick_and_place_demo')
        
        # 初始化场景对象
        scene = PlanningSceneInterface()
        
        # 创建一个发布场景变化信息的发布者
        self.scene_pub = rospy.Publisher('planning_scene', PlanningScene, queue_size=10)
        
        # 创建一个发布抓取姿态的发布者
        self.gripper_pose_pub = rospy.Publisher('gripper_pose', PoseStamped, queue_size=10)
        
        # 创建一个存储物体颜色的字典对象
        self.colors = dict()
                        
        # 初始化需要使用move group控制的机械臂中的arm group
        arm = MoveGroupCommander(GROUP_NAME_ARM)
        
        # 初始化需要使用move group控制的机械臂中的gripper group
        gripper = MoveGroupCommander(GROUP_NAME_GRIPPER)

        # 获取终端link的名称
        #end_effector_link = arm.get_end_effector_link()
 
        # 设置位置(单位：米)和姿态（单位：弧度）的允许误差
        arm.set_goal_position_tolerance(0.05)
        arm.set_goal_orientation_tolerance(0.1)

        # 当运动规划失败后，允许重新规划
        arm.allow_replanning(True)
        
        # 设置目标位置所使用的参考坐标系
        #arm.set_pose_reference_frame(REFERENCE_FRAME)
        
        # 设置每次运动规划的时间限制：20s
        arm.set_planning_time(20)
        
        # 设置pick和place阶段的最大尝试次数
        max_pick_attempts = 5
        max_place_attempts = 5
        rospy.sleep(2)

        # 设置场景物体的名称 
        table1_id = 'table1'
        table2_id = 'table2'
        target_id = 'cube_marker'
                
        # 移除场景中之前运行残留的物体
        scene.remove_world_object(table1_id)
        scene.remove_world_object(table2_id)
        scene.remove_world_object(target_id)
        
        # 移除场景中之前与机器臂绑定的物体
        scene.remove_attached_object(GRIPPER_FRAME, target_id)  
        rospy.sleep(1)
        
        # 控制机械臂先运动到准备位置
        arm.set_named_target('home')
        arm.go()
        
        # 控制夹爪张开
        gripper.set_joint_value_target(GRIPPER_OPEN)
        gripper.go()
        rospy.sleep(1)
     
        # 设置table的三维尺寸[长, 宽, 高]
        table1_size = [0.8, 1.5, 0.03]
        table2_size = [1.5, 0.8, 0.03]
        
        # 将两张桌子加入场景当中
        table1_pose = PoseStamped()
        table1_pose.header.frame_id = 'base_link'
        table1_pose.pose.position.x = 0
        table1_pose.pose.position.y = 1.05
        table1_pose.pose.position.z = -0.05
        table1_pose.pose.orientation.w = 1.0
        scene.add_box(table1_id, table1_pose, table1_size)

        table2_pose = PoseStamped()
        table2_pose.header.frame_id = 'base_link'
        table2_pose.pose.position.x = -1.05
        table2_pose.pose.position.y = -0.1
        table2_pose.pose.position.z = -0.05
        table2_pose.pose.orientation.w = 1.0
        scene.add_box(table2_id, table2_pose, table2_size)
                             
        # 将桌子设置成红色
        self.setColor(table1_id, 0.8, 0.4, 0, 1.0)
        self.setColor(table2_id, 0.8, 0.4, 0, 1.0)

        #监听目标到'base_link'的tf变换
        listener = tf.TransformListener()
        while not rospy.is_shutdown():
            try:
                (trans,rot) = listener.lookupTransform('base_link', 'ar_marker_0', rospy.Time(0))
                break
            except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
                rospy.loginfo("Waiting for transform between 'base_link' and 'ar_marker_0'")
                rospy.sleep(1)

        rospy.loginfo("Found transform between 'base_link' and 'ar_marker_0'") 

        # 设置目标物体的尺寸
        target_size = [0.05, 0.05, 0.2]
        
        # 设置目标物体的位置
        target_pose = PoseStamped()
        target_pose.header.frame_id = REFERENCE_FRAME
        target_pose.pose.position.x = trans[0]
        target_pose.pose.position.y = trans[1]
        target_pose.pose.position.z = 0.065
        target_pose.pose.orientation.x = rot[0]
        target_pose.pose.orientation.y = rot[1]
        target_pose.pose.orientation.z = rot[2]
        target_pose.pose.orientation.w = rot[3]
       
        # 将抓取的目标物体加入场景中
        scene.add_box(target_id, target_pose, target_size)
        
        # 将目标物体设置为黄色
        self.setColor(target_id, 0.9, 0.9, 0, 1.0)
        
        # 将场景中的颜色设置发布
        self.sendColors()

        # 设置支持的外观
        arm.set_support_surface_name(table2_id)
        
        # 设置一个place阶段需要放置物体的目标位置
        place_pose = PoseStamped()
        place_pose.header.frame_id = REFERENCE_FRAME
        place_pose.pose.position.x =  -0.5
        place_pose.pose.position.y =  0
        place_pose.pose.position.z =  0.065

        while not rospy.is_shutdown():
            try:
                listener.waitForTransform('ar_marker_0', 'ee_link', rospy.Time(0),rospy.Duration(4.0))
                break
            except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
                rospy.loginfo("Waiting for transform between 'ar_marker_0' and 'ee_link'")
                rospy.sleep(1)

        rospy.loginfo("Found transform between 'ar_marker_0' and 'ee_link'") 

        # 将目标位置设置为机器人的抓取目标位置
        grasp_pose = PoseStamped()
        grasp_pose.header.frame_id = 'ar_marker_0'
        grasp_pose.pose.position.x =  0
        grasp_pose.pose.position.y =  -0.15
        grasp_pose.pose.position.z =  -0.1
                
        # 生成抓取姿态
        grasps = self.make_grasps(grasp_pose, [target_id])

        # 将抓取姿态发布，可以在rviz中显示
        for grasp in grasps:
            self.gripper_pose_pub.publish(grasp.grasp_pose)
            rospy.sleep(0.2)
    
        # 追踪抓取成功与否，以及抓取的尝试次数
        result = None
        n_attempts = 0
        
        # 重复尝试抓取，直道成功或者超多最大尝试次数
        while result != MoveItErrorCodes.SUCCESS and n_attempts < max_pick_attempts:
            n_attempts += 1
            rospy.loginfo("Pick attempt: " +  str(n_attempts))
            result = arm.pick(target_id, grasps)
            rospy.sleep(0.2)
        
        # 如果pick成功，则进入place阶段 
        if result == MoveItErrorCodes.SUCCESS:
            result = None
            n_attempts = 0
            
            # 生成放置姿态
            places = self.make_places(place_pose)
            
            # 重复尝试放置，直道成功或者超多最大尝试次数
            while result != MoveItErrorCodes.SUCCESS and n_attempts < max_place_attempts:
                n_attempts += 1
                rospy.loginfo("Place attempt: " +  str(n_attempts))
                for place in places:
                    result = arm.place(target_id, place)
                    if result == MoveItErrorCodes.SUCCESS:
                        break
                rospy.sleep(0.2)
                
            if result != MoveItErrorCodes.SUCCESS:
                rospy.loginfo("Place operation failed after " + str(n_attempts) + " attempts.")
        else:
            rospy.loginfo("Pick operation failed after " + str(n_attempts) + " attempts.")
                
        # 控制机械臂回到初始化位置
        arm.set_named_target('home')
        arm.go()
        
        # 控制夹爪回到张开的状态
        gripper.set_joint_value_target(GRIPPER_OPEN)
        gripper.go()
        rospy.sleep(1)

        # 关闭并退出moveit
        moveit_commander.roscpp_shutdown()
        moveit_commander.os._exit(0)
        
    # 创建夹爪的姿态数据JointTrajectory
    def make_gripper_posture(self, joint_positions):
        # 初始化夹爪的关节运动轨迹
        t = JointTrajectory()
        
        # 设置夹爪的关节名称
        t.joint_names = ['gripper_right_driver_joint']
        
        # 初始化关节轨迹点
        tp = JointTrajectoryPoint()
        
        # 将输入的关节位置作为一个目标轨迹点
        tp.positions = joint_positions
        
        # 设置夹爪的力度
        tp.effort = [2.0]
        
        # 设置运动时间
        tp.time_from_start = rospy.Duration(1.0)
        
        # 将目标轨迹点加入到运动轨迹中
        t.points.append(tp)
        
        # 返回夹爪的关节运动轨迹
        return t
    
    # 使用给定向量创建夹爪的translation结构
    def make_gripper_translation(self, min_dist, desired, frame, vector):
        # 初始化translation对象
        g = GripperTranslation()
        
        # 设置方向向量
        g.direction.vector.x = vector[0]
        g.direction.vector.y = vector[1]
        g.direction.vector.z = vector[2]
        
        # 设置参考坐标系
        g.direction.header.frame_id = frame
        
        # 设置最小和期望的距离
        g.min_distance = min_dist
        g.desired_distance = desired
        
        return g

    # 创建一个允许的的抓取姿态列表
    def make_grasps(self, initial_pose_stamped, allowed_touch_objects):
        # 初始化抓取姿态对象
        g = Grasp()
        
        # 创建夹爪张开、闭合的姿态
        g.pre_grasp_posture = self.make_gripper_posture(GRIPPER_OPEN)

        g.grasp_posture = self.make_gripper_posture(GRIPPER_CLOSED)
                
        # 设置期望的夹爪靠近、撤离目标的参数
        g.pre_grasp_approach = self.make_gripper_translation(0.05, 0.1, 'ar_marker_0' ,[0.0, 1.0, 0.0])
        g.post_grasp_retreat = self.make_gripper_translation(0.1, 0.15,'base_link', [0.0, 0.0, 1.0])
        
        # 设置抓取姿态
        g.grasp_pose = initial_pose_stamped

        # 需要尝试改变姿态的数据列表
        yaw_vals = [0, 0.1, -0.1, 0.2, -0.2, 0.3, -0.3]
        y_vals = [0, 0.005, 0.01, 0.015, -0.005, -0.01, -0.015]

        # 抓取姿态的列表
        grasps = []

        # 改变姿态，生成抓取动作
        #for Y in yaw_vals:

        # 欧拉角到四元数的转换
        q = quaternion_from_euler(3.1415, 0, 1.5707)                
        # 设置抓取的姿态
        g.grasp_pose.pose.orientation.x = q[0]
        g.grasp_pose.pose.orientation.y = q[1]
        g.grasp_pose.pose.orientation.z = q[2]
        g.grasp_pose.pose.orientation.w = q[3]
        
        # 设置抓取的唯一id号
        g.id = str(len(grasps))
        
        # 设置允许接触的物体
        g.allowed_touch_objects = allowed_touch_objects
        
        # 将本次规划的抓取放入抓取列表中
        grasps.append(deepcopy(g))
       
        # 返回抓取列表
        return grasps
    
    # 创建一个允许的放置姿态列表
    def make_places(self, init_pose):
        # 初始化放置抓取物体的位置
        place = PoseStamped()
        
        # 设置放置抓取物体的位置
        place = init_pose
        
        # 定义x方向上用于尝试放置物体的偏移参数
        x_vals = [0, 0.005, 0.01, 0.015, -0.005, -0.01, -0.015]
        
        # 定义y方向上用于尝试放置物体的偏移参数
        y_vals = [0, 0.005, 0.01, 0.015, -0.005, -0.01, -0.015]
        
        pitch_vals = [0]
        
        # 定义用于尝试放置物体的偏航角参数
        yaw_vals = [0]

        # 定义放置物体的姿态列表
        places = []
        
        # 生成每一个角度和偏移方向上的抓取姿态
        #for Y in yaw_vals:
           # for p in pitch_vals:
        for y in y_vals:
            for x in x_vals:
                place.pose.position.x = init_pose.pose.position.x + x
                place.pose.position.y = init_pose.pose.position.y + y
                
                # 欧拉角到四元数的转换
                q = quaternion_from_euler(3.1415, 0, 3.1415)
                
                # 欧拉角到四元数的转换
                place.pose.orientation.x = q[0]
                place.pose.orientation.y = q[1]
                place.pose.orientation.z = q[2]
                place.pose.orientation.w = q[3]
                
                # 将该放置姿态加入列表
                places.append(deepcopy(place))
        
        # 返回放置物体的姿态列表
        return places
    
    # 设置场景物体的颜色
    def setColor(self, name, r, g, b, a = 0.9):
        # 初始化moveit颜色对象
        color = ObjectColor()
        
        # 设置颜色值
        color.id = name
        color.color.r = r
        color.color.g = g
        color.color.b = b
        color.color.a = a
        
        # 更新颜色字典
        self.colors[name] = color

    # 将颜色设置发送并应用到moveit场景当中
    def sendColors(self):
        # 初始化规划场景对象
        p = PlanningScene()

        # 需要设置规划场景是否有差异     
        p.is_diff = True
        
        # 从颜色字典中取出颜色设置
        for color in self.colors.values():
            p.object_colors.append(color)
        
        # 发布场景物体颜色设置
        self.scene_pub.publish(p)

if __name__ == "__main__":
    PickAndPlaceDemo()

    